File: kgg.py
#!/usr/bin/env python3
"""酷狗新版 .kgg 解密的密钥库（KGMusicV3.db）定位。
新版 kgg 的密钥不在文件里，存在酷狗 PC 客户端（win32 v11）的私有数据库
KGMusicV3.db 中（与 QQ musicex 把 ekey 锁进 mmkv 同一套路）。
本模块负责在常见位置自动找库，并记住找到（或用户最后兜底选定）的路径。"""
import os
import sys

if os.name == "nt":
    MU_DIR = os.path.join(
        os.environ.get("LOCALAPPDATA", os.path.expanduser(r"~\AppData\Local")),
        "music-unlock")
else:
    MU_DIR = os.path.expanduser("~/.local/share/music-unlock")
CONF = os.path.join(MU_DIR, "kgg_db_path.txt")
DB_NAME = "KGMusicV3.db"

def default_roots():
    """Windows 只找酷狗自己的目录；不要扫整个 AppData（微信/浏览器会把扫描拖死）。"""
    if os.name == "nt":
        roaming = os.environ.get("APPDATA") or os.path.expanduser(r"~\AppData\Roaming")
        local = os.environ.get("LOCALAPPDATA") or os.path.expanduser(r"~\AppData\Local")
        return [
            os.path.join(roaming, "KuGou8"),
            os.path.join(roaming, "Kugou8"),
            os.path.join(roaming, "KuGou"),
            os.path.join(local, "KuGou8"),
            os.path.join(local, "KuGou"),
        ]
    return [
        "~/.cxoffice",
        "~/.wine",
        "~/.local/share/wineprefixes",
        "~/Games",
        "~/Applications",
    ]


ROOTS = default_roots()

SKIP_DIRS = {
    "windows", "system32", "syswow64", "winsxs", "node_modules",
    ".git", "$recycle.bin", "temp", "tmp",
    "steam", "steamapps", ".steam",
}


def _remember(path):
    os.makedirs(MU_DIR, exist_ok=True)
    with open(CONF, "w", encoding="utf-8") as f:
        f.write(path)
    return path


def saved_db():
    try:
        with open(CONF, encoding="utf-8") as f:
            p = f.read().strip()
        return p if p and os.path.exists(p) else None
    except OSError:
        return None


def set_db(path):
    return _remember(path)


def usable_db(path):
    """Windows 上酷狗开着会锁 KGMusicV3.db，拷一份给 um 用。"""
    import win32compat
    if sys.platform != "win32":
        return path
    return win32compat.snapshot_sqlite(path, os.path.join(MU_DIR, "kgg-snap"))


def find_db():
    """自动搜 KGMusicV3.db：先 remembered，再扫常见根目录（限深防慢）。"""
    p = saved_db()
    if p:
        return p
    for root in ROOTS:
        root = os.path.expanduser(root)
        if not os.path.isdir(root):
            continue
        for dirpath, dirnames, filenames in os.walk(root):
            depth = dirpath[len(root):].count(os.sep)
            dirnames[:] = [d for d in dirnames if d.lower() not in SKIP_DIRS]
            if depth > 6:
                dirnames[:] = []
                continue
            if DB_NAME in filenames:
                return _remember(os.path.join(dirpath, DB_NAME))
    return None

File: crates/core/src/kgg.rs
//! 酷狗 KGMusicV3.db 定位，对照 kgg.py。Windows 只扫酷狗自己的目录。

use std::fs;
use std::path::{Path, PathBuf};

pub const DB_NAME: &str = "KGMusicV3.db";

const SKIP: &[&str] = &[
    "windows", "system32", "syswow64", "winsxs", "node_modules", ".git", "$recycle.bin", "temp",
    "tmp", "steam", "steamapps", ".steam",
];

#[derive(Clone, Debug)]
pub struct KggPaths {
    pub mu_dir: PathBuf,
    pub roots: Vec<PathBuf>,
}

impl KggPaths {
    pub fn conf(&self) -> PathBuf {
        self.mu_dir.join("kgg_db_path.txt")
    }
}

pub fn default_roots() -> Vec<PathBuf> {
    if cfg!(windows) {
        let roaming = std::env::var_os("APPDATA")
            .map(PathBuf::from)
            .unwrap_or_else(|| dirs_fallback("AppData/Roaming"));
        let local = std::env::var_os("LOCALAPPDATA")
            .map(PathBuf::from)
            .unwrap_or_else(|| dirs_fallback("AppData/Local"));
        ["KuGou8", "Kugou8", "KuGou"]
            .into_iter()
            .flat_map(|name| [roaming.join(name), local.join(name)])
            .collect()
    } else {
        ["~/.cxoffice", "~/.wine", "~/.local/share/wineprefixes", "~/Games", "~/Applications"]
            .into_iter()
            .map(expand_home)
            .collect()
    }
}

fn dirs_fallback(rest: &str) -> PathBuf {
    expand_home("~").join(rest)
}

fn expand_home(p: &str) -> PathBuf {
    if let Some(rest) = p.strip_prefix("~/") {
        let home = std::env::var_os("HOME")
            .or_else(|| std::env::var_os("USERPROFILE"))
            .map(PathBuf::from)
            .unwrap_or_else(|| PathBuf::from("."));
        return home.join(rest);
    }
    PathBuf::from(p)
}

fn remember(conf: &Path, path: &Path) -> PathBuf {
    if let Some(parent) = conf.parent() {
        let _ = fs::create_dir_all(parent);
    }
    let _ = fs::write(conf, path.to_string_lossy().as_bytes());
    path.to_path_buf()
}

pub fn find_db(paths: &KggPaths) -> Option<PathBuf> {
    let conf = paths.conf();
    if let Ok(saved) = fs::read_to_string(&conf) {
        let p = PathBuf::from(saved.trim());
        if p.is_file() {
            return Some(p);
        }
    }
    for root in &paths.roots {
        if !root.is_dir() {
            continue;
        }
        if let Some(found) = walk(root, 0) {
            return Some(remember(&conf, &found));
        }
    }
    None
}

fn walk(dir: &Path, depth: usize) -> Option<PathBuf> {
    let rd = fs::read_dir(dir).ok()?;
    let mut dirs = Vec::new();
    for ent in rd.flatten() {
        let name = ent.file_name();
        let name_s = name.to_string_lossy();
        let path = ent.path();
        if path.is_file() && name_s == DB_NAME {
            return Some(path);
        }
        if path.is_dir() && !SKIP.iter().any(|s| name_s.eq_ignore_ascii_case(s)) {
            dirs.push(path);
        }
    }
    if depth >= 6 {
        return None;
    }
    for d in dirs {
        if let Some(found) = walk(&d, depth + 1) {
            return Some(found);
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn first_scan_returns_the_path() {
        let tmp = tempdir().unwrap();
        let nested = tmp.path().join("KuGou8");
        fs::create_dir(&nested).unwrap();
        let db = nested.join(DB_NAME);
        fs::write(&db, b"x").unwrap();
        let confdir = tmp.path().join("conf");
        let found = find_db(&KggPaths {
            mu_dir: confdir.clone(),
            roots: vec![tmp.path().to_path_buf()],
        })
        .unwrap();
        assert_eq!(found, db);
        assert_eq!(fs::read_to_string(confdir.join("kgg_db_path.txt")).unwrap(), db.to_string_lossy());
    }

    #[test]
    fn missing_db_returns_none() {
        let tmp = tempdir().unwrap();
        assert!(find_db(&KggPaths {
            mu_dir: tmp.path().join("conf"),
            roots: vec![tmp.path().to_path_buf()],
        })
        .is_none());
    }

    #[test]
    fn windows_roots_are_kugou_not_whole_appdata() {
        // 逻辑对照 kgg.default_roots：Windows 条目都以酷狗目录结尾。
        let roots = default_roots();
        if cfg!(windows) {
            assert!(roots.iter().all(|p| {
                p.ends_with("KuGou8") || p.ends_with("Kugou8") || p.ends_with("KuGou")
            }));
        } else {
            assert!(roots.iter().any(|p| p.ends_with("wine") || p.ends_with(".wine")));
        }
    }
}
